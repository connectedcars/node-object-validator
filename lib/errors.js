function ValidationError(message, context = {}) {
  this.name = 'ValidationError'
  this.message = message
  if (context.key) {
    this.key = context.key
  }
  if (context.val) {
    this.val = context.val
  }
}
ValidationError.prototype = new Error()

module.exports = {
  ValidationError
}
