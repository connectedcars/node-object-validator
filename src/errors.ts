export interface ValidationErrorContext<T> {
  key: string
  value: T
}

export class ValidationError<T> extends Error {
  public key?: string
  public value?: T
  public constructor(message: string, context?: ValidationErrorContext<T>) {
    if (context) {
      message = `Field '${context.key}' ${message.charAt(0).toLowerCase() + message.slice(1)}`
    }
    super(message)
    this.key = context?.key
    this.value = context?.value
  }
}

export class RequiredError extends ValidationError<string> {}
export class NotStringError extends ValidationError<string> {}
export class WrongLengthError extends ValidationError<string> {}
export class NotDateError extends ValidationError<string> {}
export class NotRfc3339Error extends ValidationError<string> {}
export class NotExactStringError extends ValidationError<string> {}
export class NotFloatError extends ValidationError<string> {}
export class NotIntegerError extends ValidationError<string> {}
export class NotArrayError extends ValidationError<string> {}
export class NotObjectError extends ValidationError<string> {}
export class DoesNotMatchRegexError extends ValidationError<string> {}
export class OutOfRangeError extends ValidationError<string> {}
