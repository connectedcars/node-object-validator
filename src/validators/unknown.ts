import { CodeGenResult, ValidatorBase, ValidatorBaseOptions, ValidatorExportOptions, ValidatorOptions } from '../common'
import { RequiredFail, ValidationFailure } from '../errors'

export abstract class UnknownValidator<O = never> extends ValidatorBase<unknown | O> {
  public constructor(options?: ValidatorBaseOptions) {
    super({ ...options })
    if (options?.optimize !== false) {
      this.optimize()
    }
  }

  public codeGen(
    valueRef: string,
    validatorRef: string,
    id = () => {
      return this.codeGenId++
    },
    context?: string,
    earlyFail?: boolean
  ): CodeGenResult {
    const contextStr = context ? `, \`${context}\`` : ', context'
    const localValueRef = `value${id()}`
    const declarations: string[] = []
    // prettier-ignore
    const code: string[] = [
      `const ${localValueRef} = ${valueRef}`,
      ...(this.required ? [
      `if (${localValueRef} === undefined) {`,
      `  errors.push(new RequiredFail(\`Is required\`, ${localValueRef}${contextStr}))`,
      '}'
      ] : []),
      ...(earlyFail ? [
      `if (errors.length > 0) {`,
      `  return errors`,
      `}`] : []),

    ]
    return [
      {
        RequiredFail: RequiredFail
      },
      declarations,
      code
    ]
  }

  public toString(options?: ValidatorExportOptions): string {
    if (options?.types === true) {
      return this.typeString(options)
    } else {
      return this.constructorString()
    }
  }

  protected validateValue(value: unknown, context?: string): ValidationFailure[] {
    if (value === undefined) {
      return this.required ? [new RequiredFail(`Is required`, value, context)] : []
    }
    return []
  }

  private typeString(options?: ValidatorExportOptions): string {
    const language = options?.language ?? 'typescript'
    switch (language) {
      case 'typescript': {
        let typeStr = `unknown`

        if (this.required === false) {
          typeStr += ` | undefined`
        }
        if (this.nullable === true) {
          typeStr += ` | null`
        }

        return typeStr
      }
      case 'rust': {
        throw new Error('Rust not supported yet')
      }
      default: {
        throw new Error(`Language: '${options?.language}' unknown`)
      }
    }
  }

  private constructorString(): string {
    return `new ${this.constructor.name}(${this.optionsString})`
  }
}

export class RequiredUnknown extends UnknownValidator {
  public constructor(options?: ValidatorOptions) {
    super({ ...options, required: true })
  }
}

export class OptionalUnknown extends UnknownValidator<undefined> {
  public constructor(options?: ValidatorOptions) {
    super({ ...options, required: false })
  }
}
