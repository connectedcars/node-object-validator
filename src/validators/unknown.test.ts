import { RequiredFail } from '../errors'
import { OptionalUnknown, RequiredUnknown, UnknownValidator } from './unknown'

describe.each([false, true])('Unknown (optimize: %s)', optimize => {
  describe('Unknown', () => {
    it('should generate validation code and give same result', () => {
      const validator = new UnknownValidator({ optimize })
      if (optimize) {
        expect(validator['optimizedValidate']).not.toBeNull()
      } else {
        expect(validator['optimizedValidate']).toBeNull()
      }
      expect(validator.validate('Anything')).toEqual([])
      expect(validator.validate(1)).toEqual([])
      expect(validator.validate(true)).toEqual([])
      expect(validator.validate([])).toEqual([])
      expect(validator.validate({})).toEqual([])
    })

    describe('RequiredUnknown', () => {
      it('requires empty value', () => {
        const validator = new RequiredUnknown()
        expect(validator.validate(null)).toStrictEqual([new RequiredFail('Is required')])
        expect(validator.validate(undefined)).toStrictEqual([new RequiredFail('Is required')])
      })
    })

    describe('OptionalString', () => {
      it('requires empty value', () => {
        const validator = new OptionalUnknown()
        expect(validator.validate(undefined)).toStrictEqual([])
        expect(validator.validate(undefined)).toStrictEqual([])
      })
    })
  })
})
