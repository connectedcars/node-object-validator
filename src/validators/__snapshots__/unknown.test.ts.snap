// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unknown RequiredUnknown should return a function body 1`] = `
Array [
  Object {
    "RequiredFail": [Function],
  },
  Array [],
  Array [
    "const value1 = value1",
    "if (value1 === undefined) {",
    "  errors.push(new RequiredFail(\`Is required\`, value1, context))",
    "}",
  ],
]
`;
